src/middlewared/middlewared/plugins/auth.py

```diff
@@ -325,9 +349,130 @@ def token(self, app, token):
+
+    class Config:
+        datastore = 'system.twofactorauthentication'
+        datastore_extend = 'auth.twofactor.two_factor_extend'
+        namespace = 'auth.twofactor'
+
+    @private
+    async def two_factor_extend(self, data):
+        data['secret'] = await self.middleware.call('pwenc.decrypt', data['secret'])
+
+        for srv in ['ssh']:
+            data['services'].setdefault(srv, False)
+
+        return data
+
+    @accepts(
+        Dict(
+            'auth_twofactor_update',
+            Bool('enabled'),
+            Int('otp_digits', validators=Range(min=6, max=8)),
+            Int('window', validators=Range(min=0)),
+            Int('interval', validators=Range(min=5)),
+            Dict(
+                'services',
+                Bool('ssh', default=False)
+            ),
+            update=True
+        )
+    )
+    async def do_update(self, data):
+        """
+        `otp_digits` represents number of allowed digits in the OTP.
+
+        `window` extends the validity to `window` many counter ticks before and after the current one.
+
+        `interval` is time duration in seconds specifying OTP expiration time from it's creation time.
+        """
+        old_config = await self.config()
+        config = old_config.copy()
+
+        config.update(data)
+
+        if config['enabled'] and not config['secret']:
+            # Only generate a new secret on `enabled` when `secret` is not already set.
+            # This will aid users not setting secret up again on their mobiles.
+            config['secret'] = await self.middleware.run_in_thread(
+                self.generate_base32_secret
+            )
+
+        config['secret'] = await self.middleware.call('pwenc.encrypt', config['secret'])
+
+        await self.middleware.call(
+            'datastore.update',
+            self._config.datastore,
+            config['id'],
+            config
+        )
+
+        await self.middleware.call('service.reload', 'ssh')
+
+        return await self.config()
+
+    @accepts(
+        Str('token', null=True)
+    )
+    def verify(self, token):
+        """
+        Returns boolean true if provided `token` is successfully authenticated.
+        """
+        config = self.middleware.call_sync(f'{self._config.namespace}.config')
+        if not config['enabled']:
+            raise CallError('Please enable Two Factor Authentication first.')
+
+        totp = pyotp.totp.TOTP(
+            config['secret'], interval=config['interval'], digits=config['otp_digits']
+        )
+        return totp.verify(token, valid_window=config['window'])
+
+    @accepts()
+    def renew_secret(self):
+        """
+        Generates a new secret for Two Factor Authentication. Returns boolean true on success.
+        """
+        config = self.middleware.call_sync(f'{self._config.namespace}.config')
+        if not config['enabled']:
+            raise CallError('Please enable Two Factor Authentication first.')
```

Comment:
```suggestion
            raise CallError('Please enable Two-Factor Authentication first.')
```
Would it be useful to the user to point them to where they can enable two-factor auth?